### DISCUSSION TITLE
[Chapter 09] 메타 프로그래밍?

### DISCUSSION THREAD

**Discussion by Irisation23** [cowaine/is-my-code-that-strange#77] [Chapter 09-10 (악마, 이름 설계)] [answered]
p200 ~ p204 참조

- 프로그램 구조 자체를 제어하는 프로그래밍을 메타 프로그래밍이라고 부름
- 자바에서 메타 프로그래밍을 활용해 클래스 구조를 읽고 쓸 때는 `Reflection API`를 사용함

하지만 Reflection은 객체지향 프로그래밍과는 거리가 있음
- 클래스, 메서드, 필드를 직접 접근하는 이유 때문에 `추상화`를 깨버림

```
p204
메타 프로그래밍을 사용해야 한다면 시스템 분석 용도로 한정하거나, 아주 작은 범위에서만 활용하는 등 리스크를 최소화해야 합니다.
```

### 질문

1. 메타 프로그래밍을 시스템 분석 용도로 사용하는 이유는 무엇일까요?
2. 메타 프로그래밍이 불러올 리스크의 예는 어떤 것이 있을까요?

**Comment by snoopingh** [comment]
1. APM 도구를 사용해보면 운영중인 애플리케이션의 API 호출 및 그에 대한 로그, 트랜잭션, 트래픽 분석 등등..을 수집해서 지표로 보여주지 않습니까? 그 때, 클래스명까지 나오던데 이것과 연관이 있을 것 같습니다.
요청 처리 시간이 지연되는 경우 어느 부분에서 지연이 생겼는지 확인하기에도 용이할 것 같습니다

2. 리스크는 저도 내공이 부족해서 잘 모르겠네요.. 리플렉션을 남발하면 findbugs나 소나큐브 같은 품질 검사 도구에서 private 필드에 접근하는 부분에 대해 경고를 많이 해줘서 수정건이 많이 생기는 것 정도가 있겠습니다

**Comment by firewoody237** [comment, marked as answer]
1 시스템 분석이란게, 시스템의 런타임에서 객체를 조작하여 시스템이 잘 돌아가는 것을 확인하거나, Proxy처럼 로그를 집어넣어 확인하는 방법들로 언급한게 아닐까 싶습니다.
![image](https://github.com/user-attachments/assets/75182907-c903-47b3-a0a2-14713ac337c7)
![image](https://github.com/user-attachments/assets/26276a24-a5bc-4f09-87b5-8863bb39e496)

2 런타임 에러 자체가 리스크 같습니다.
리플렉션같은 경우 런타임에 어떤 예외가 발생할지 추측하기도 어렵고, 런타임 에러 자체가 IDE를 통해 사전 처리하지 못하는 예외이기 때문에 리스크 같아요.
